### PR TITLE
[RW-4736][risk=no] Puppeteer test: waitFor institution email validation

### PR DESCRIPTION
**Pull request by aweng98**
Changes will also resolve ticket RW-4648.

### CODE REVIEW

File: e2e/app/aou-elements/clr-icon-link.ts [hunk 1]
```diff
@@ -1,23 +1,24 @@
...
-     aElementName: string,
+     textOptions: TextOptions,
      shape: string,
      waitOptions: WaitForSelectorOptions = {visible: true},
      throwErr = true): Promise<ClrIconLink> {
 
     let element: ClrIconLink;
     try {
-      const iconElement = await findIcon(page, aElementName, shape, waitOptions);
+      const iconElement = await findIcon(page, textOptions, shape, waitOptions);
```
Comment (jaycarlton) [inline e2e/app/aou-elements/clr-icon-link.ts:17]: Do you need to catch and rethrow, or will the framework & language do most of this for you?

File: e2e/app/aou-elements/xpath-defaults.ts [hunk 2]
```diff
@@ -2,11 +2,11 @@ import TextOptions from './text-options';
 
 function textXpathHelper(opts: TextOptions) {
   if (opts.text) {
-    return `text()="${opts.text}" or @aria-label="${opts.text}" or @placeholder="${opts.text}"`;
+    return `text()='${opts.text}' or @aria-label='${opts.text}' or @placeholder='${opts.text}'`;
   } else if (opts.textContains) {
-    return `contains(text(),"${opts.textContains}") or contains(@aria-label,"${opts.textContains}") or contains(@placeholder,"${opts.textContains}")`;
+    return `contains(text(),'${opts.textContains}') or contains(@aria-label,'${opts.textContains}') or contains(@placeholder,'${opts.textContains}')`;
```
Comment (jaycarlton) [inline e2e/app/aou-elements/xpath-defaults.ts:7]: nit: space after commas

File: e2e/app/aou-elements/xpath-defaults.ts [hunk 3]
```diff
@@ -2,11 +2,11 @@ import TextOptions from './text-options';
 
 function textXpathHelper(opts: TextOptions) {
   if (opts.text) {
-    return `text()="${opts.text}" or @aria-label="${opts.text}" or @placeholder="${opts.text}"`;
+    return `text()='${opts.text}' or @aria-label='${opts.text}' or @placeholder='${opts.text}'`;
```
Comment (jaycarlton) [inline e2e/app/aou-elements/xpath-defaults.ts:5]: nit: can you declare a string for `'${opts.text}'` and use it in all three places?

File: e2e/app/aou-elements/xpath-defaults.ts [hunk 4]
```diff
@@ -59,10 +59,11 @@ export function clickableXpath(label: string) {
  * @param label:
  * @param shapeValue:
  */
-export function clrIconXpath(label: string, shapeValue: string) {
-  if (label === '') {
+export function clrIconXpath(opts: TextOptions, shapeValue: string) {
```
Comment (jaycarlton) [inline e2e/app/aou-elements/xpath-defaults.ts:62]: Is xpath the only way to reach this icon? I feel like it may be brittle in the face of future page reorganization and isn't the most readable representation for folks trying to investigate failed tests.

Also, I don't know what clr means.

File: e2e/app/authenticated-page.ts [hunk 5]
```diff
@@ -132,7 +132,7 @@ export default abstract class AuthenticatedPage extends BasePage {
     const is = await this.isSideNavDropdownOpen();
     if (!is) {
       // click bars icon to open dropdown
-      const icon = await findIcon(this.page, '', 'bars');
+      const icon = await findIcon(this.page, {}, 'bars');
```
Comment (jaycarlton) [inline e2e/app/authenticated-page.ts:135]: Can you name this `barsIcon` or `hamburgerIcon`?

File: e2e/app/authenticated-page.ts [hunk 6]
```diff
@@ -132,7 +132,7 @@ export default abstract class AuthenticatedPage extends BasePage {
     const is = await this.isSideNavDropdownOpen();
```
Comment (jaycarlton) [inline e2e/app/authenticated-page.ts:132]: nit: name this `sideNavIsOpen`

File: e2e/app/authenticated-page.ts [hunk 7]
```diff
@@ -179,7 +179,7 @@ export default abstract class AuthenticatedPage extends BasePage {
 
   private async isSideNavDropdownOpen(): Promise<boolean> {
```
Comment (jaycarlton) [inline e2e/app/authenticated-page.ts:180]: Could you please add a description, like "the side Nav for this page is either docked to the right or open. We can determine the open state by looking for a visible Home icon" Or something. It's not obvious to someone who doesn't watch the UI closely.

Additionally, it's confusing that the method is named after the  side nav dropdown, but the implementation is in terms of a  home icon. Can we add a test data tag to the nav and check for it more directly?

File: e2e/app/create-account-page.ts [hunk 8]
```diff
@@ -154,7 +154,7 @@ export default class CreateAccountPage extends BasePage {
     const emailAddress = await Textbox.forLabel(this.page, {textContains: FIELD_LABEL.INSTITUTION_EMAIL, ancestorNodeLevel: 2});
```
Comment (jaycarlton) [inline e2e/app/create-account-page.ts:154]: I would pass an institution input object into this method so that you can fill it out more than one way to test edge cases, validation, etc, if you want.

File: e2e/app/home-page.ts [hunk 9]
```diff
@@ -32,7 +32,7 @@ export default class HomePage extends AuthenticatedPage {
   }
 
   async getCreateNewWorkspaceLink(): Promise<ElementHandle> {
-    return findIcon(this.page, FIELD_LABEL.CREATE_NEW_WORKSPACE, 'plus-circle');
+    return findIcon(this.page, {text: FIELD_LABEL.CREATE_NEW_WORKSPACE}, 'plus-circle');
```
Comment (jaycarlton) [inline e2e/app/home-page.ts:35]: nit: could we make an enum  with these icon type strings?

File: e2e/tests/user/registration-ui.spec.ts [hunk 10]
```diff
@@ -1,7 +1,5 @@
 import BaseElement from '../../app/aou-elements/base-element';
-import SelectComponent from '../../app/aou-elements/select-component';
-import Textbox from '../../app/aou-elements/textbox';
-import CreateAccountPage, {FIELD_LABEL, INSTITUTION_ROLE_VALUE, INSTITUTION_VALUE} from '../../app/create-account-page';
+import CreateAccountPage from '../../app/create-account-page';
```
Comment (jaycarlton) [inline e2e/tests/user/registration-ui.spec.ts:2]: Can we use module exports or something so we don't have to import from grandparent directories? It's really breakable when you do that, because it means you can't move your test directories around easily.

File: e2e/tests/user/registration-ui.spec.ts [hunk 11]
```diff
@@ -90,14 +88,7 @@ describe('User registration tests:', () => {
...
-    await institutionSelect.select(INSTITUTION_VALUE.BROAD);
-
-    const emailAddress = await Textbox.forLabel(page, {textContains: FIELD_LABEL.INSTITUTION_EMAIL, ancestorNodeLevel: 2});
-    await emailAddress.type(configs.broadInstitutionEmail);
-
-    const roleSelect = new SelectComponent(page, 'describes your role');
-    await roleSelect.select(INSTITUTION_ROLE_VALUE.UNDERGRADUATE_STUDENT);
+    await createAccountPage.fillOutInstitution();
 
     await nextButton.waitUntilEnabled();
```
Comment (jaycarlton) [inline e2e/tests/user/registration-ui.spec.ts:93]: nit: there are two different conventions on this line and the next for async functions. I would have expected to see `await nextButton.enabled();` for consistency.

File: e2e/app/create-account-page.ts (hunk 8 above)
Comment (jaycarlton) [inline e2e/app/create-account-page.ts:154]: nit: I'd rename `emailAddress` to `emailAddressTextBox` to distinguish it from the address itself.

File: e2e/app/authenticated-page.ts [hunk 12]
```diff
@@ -179,7 +179,7 @@ export default abstract class AuthenticatedPage extends BasePage {
...
     try {
-      await findIcon(this.page, 'Home', 'home', {visible: true, timeout: 1000});
+      await findIcon(this.page, {text: 'Home'}, 'home', {visible: true, timeout: 1000});
```
Comment (jaycarlton) [inline e2e/app/authenticated-page.ts:182]: can you make a utility method that's an async boolean-valued method to wait look for and wait for an icon (or other element)? It seems like this would be a common itch to scratch.

File: e2e/app/authenticated-page.ts (hunk 7 above)
Comment (aweng98) [reply e2e/app/authenticated-page.ts:180]: sorry. your feedback is awesome! I have comments for most "big" methods but don't do for all methods because I think some methods are easy to know what it suppose to do. For this method, I added what you said as comment. 

It is the easiest way to determine the open or close state by looking for visible `Home` icon. If someone is going to implement a test friendly tag in future, I'd be happy to update this code.

File: e2e/app/create-account-page.ts (hunk 8 above)
Comment (aweng98) [reply e2e/app/create-account-page.ts:154]: I can refactor in a new PR. This PR only deals with one test failure that only happens on CircleCi.

File: e2e/app/home-page.ts (hunk 9 above)
Comment (aweng98) [reply e2e/app/home-page.ts:35]: good suggestion. I can refactor code in a new PR.

File: e2e/tests/user/registration-ui.spec.ts (hunk 10 above)
Comment (aweng98) [reply e2e/tests/user/registration-ui.spec.ts:2]: don't know how to do what you suggested. but I will research around for how-to to see what other people are doing.

File: e2e/app/aou-elements/button.ts [hunk 13]
```diff
@@ -27,12 +27,13 @@ export default class Button extends BaseElement {
...
-    const handle = this.element.asElement();
-    return await handle.evaluateHandle((e) => {
+  async waitUntilEnabled(): Promise<unknown> {
+    const result = await this.page.waitForFunction((e) => {
       const style = window.getComputedStyle(e);
       return style.getPropertyValue('cursor') === 'pointer';
-    }, this.element);
+    }, {}, this.element);
+    const jval = await result.jsonValue();
+    return jval;
```
Comment (aweng98) [inline e2e/app/aou-elements/button.ts:36]: Request `GET /v1/institutions/Broad/checkEmail/aou-dev-registration@broadinstitute.org/` could takes up to several seconds to complete. Changed to `waitForFunction` for waiting function returns true, up to 30 seconds.

File: e2e/app/aou-elements/clr-icon-link.ts (hunk 1 above)
Comment (aweng98) [reply e2e/app/aou-elements/clr-icon-link.ts:17]: `jest` will throws exception too. I'm printing out `textOptions` for manual checking if fail

File: e2e/app/aou-elements/button.ts [hunk 14]
```diff
@@ -27,12 +27,13 @@ export default class Button extends BaseElement {
   /**
    * Wait until button is clickable (enabled).
    */
-  async waitUntilEnabled(): Promise<JSHandle> {
-    const handle = this.element.asElement();
-    return await handle.evaluateHandle((e) => {
+  async waitUntilEnabled(): Promise<unknown> {
+    const result = await this.page.waitForFunction((e) => {
```
Comment (blrubenstein) [inline e2e/app/aou-elements/button.ts:31]: Nit: Could you use a more descriptive variable name than `result`?

File: e2e/app/aou-elements/button.ts [hunk 15]
```diff
@@ -27,12 +27,13 @@ export default class Button extends BaseElement {
...
       const style = window.getComputedStyle(e);
       return style.getPropertyValue('cursor') === 'pointer';
-    }, this.element);
+    }, {}, this.element);
+    const jval = await result.jsonValue();
```
Comment (blrubenstein) [inline e2e/app/aou-elements/button.ts:35]: Nit: I would leave out this variable, and just `return await result.jsonValue();`

If you do need the variable, I would use a more descriptive name than jval. Maybe `{rename of result}AsJson`?

File: e2e/app/aou-elements/xpath-defaults.ts [hunk 16]
```diff
@@ -1,12 +1,16 @@
 import TextOptions from './text-options';
 
 function textXpathHelper(opts: TextOptions) {
+  let txt;
```
Comment (blrubenstein) [inline e2e/app/aou-elements/xpath-defaults.ts:4]: Nit: I'd just call it `text`. Or if we can have a more descriptive name describing where the text comes from? Because I am not super following what text this variable is meant to contain.

File: e2e/tests/user/registration-ui.spec.ts (hunk 10 above)
Comment (blrubenstein) [reply e2e/tests/user/registration-ui.spec.ts:2]: I think it should be possible to specify a root directory, so you can use absolute paths from the root rather than relative from here.

File: e2e/app/aou-elements/button.ts (hunk 14 above)
Comment (aweng98) [reply e2e/app/aou-elements/button.ts:31]: how about `retrievedValue`?

File: e2e/app/aou-elements/button.ts (hunk 15 above)
Comment (aweng98) [reply e2e/app/aou-elements/button.ts:35]: done

File: e2e/app/aou-elements/xpath-defaults.ts (hunk 16 above)
Comment (aweng98) [reply e2e/app/aou-elements/xpath-defaults.ts:4]: how about `findText`?
